salt/modules/file.py

```diff
@@ -3795,6 +3809,17 @@ def is_link(path):
     # therefore a custom function will need to be called. This function
     # therefore helps API consistency by providing a single function to call for
     # both operating systems.
+    if nostat:
+        if not os.path.exists(path):
+            return False
+
+        parent_directory = os.path.dirname(path)
+
+        with os.scandir(path=parent_directory) as directory_contents:
+            for item in directory_contents:
+                if item.path == path:
+                    return item.is_symlink()
```

Comment:
Why get the contents of the whole directory when you have the path you want?!
```python
>>> import os.path
>>> base = os.path.dirname("requirements/static/ci/common.in")
>>> name = os.path.basename("requirements/static/ci/common.in")
>>> base
'requirements/static/ci'
>>> name
'common.in'
>>> import pathlib
>>> pathlib.Path(base, name).is_symlink()
False
```
Or even
```
pathlib.Path("requirements/static/ci/common.in").is_symlink()
```

I might not be understanding the actual problem being solved....

What is a hung mount?

Reply:
@Kickball can probably describe the scenario better than me, but I believe looking at the parent directory contents prevents lstat from blocking when you attempt direct access to the file. The parent directory will probably just look empty and skip.

Reply:
> directory will probably just look empty and skip

What if it has 1 file and the broken symlink?

What if it has 3000 files and the broken symlink is the last entry?

Reply:
Hi Pedro - I wanted to reply to provide some more context on this change and some performance testing figures.

Let me know if anything is unclear

**NFS Hangs**

A hung mount refers to an [hard](https://serverfault.com/questions/9499/what-are-the-advantages-disadvantages-of-hard-versus-soft-mounts-in-unix) NFS mount which has become unresponsive (e.g. filer upgrades / failures, network misconfigurations/etc) or blocked (e.g. firewall rules) causing which read/write operations on the mountpoint and its  subdirectories to hang (process in uninterruptible sleep/D state) indefinetly until unmounted or it becomes responsive again.

I performed additional testing using nftables rules to drop all traffic to a mount NFS filer to simulate this event.

* /mnt/mountpoint
  * Method 1: False
  * Method 2: Hung
  * Method 3: Hung
* /mnt/mountpoint/subdirectory
  * Method 1: Hung
  * Method 2: Hung
  * Method 3: Hung
 
Method 1 performed better in this scenario, as it was able to determine whether a hung mountpoint (e.g. /mnt/nfsserver) was a symlink without getting stuck stating the mountpoint, which hung for methods 2/3 as described above. However all 3 methods hung when checking a subdirectory on the mount, as even the parent directory would be on NFS.

This shows that the change proposed here solves the problem of hung mounts when checking the mountpoints themselves, while it doesn't solve checking subdirectories.


**Testing Context**
Method explanation:
- Method 1 is `is_link(file, nostat=True)`
- Method 2 is `is_link(file, nostat=False)`
- Method 3 is `pathlib.Path(file).is_symlink()`

Test targets:
- `/usr/directory` only contains the files which are tested (e.g. target, symlink, broken_symlink)
- `/usr/bigdirectory` contains 3000 x 15MB files, named 1 through 3000 (meaning the files we are checking will be sorted after these)
- `/mnt/mountpoint` is a remote NFSv3 file server 

Misc:
- The tests cover the concerns you mentioned above (broken symlinks and 3000 files prior to the target).
- Timings are in nanoseconds (e.g. 1 / hundredth million of a second) captured by `time.perf_counter_ns()`.
- Each file/method combination is tested 100 times on a virtual machine (so raw numbers are not very useful).

Results:
- Method 1 is orders of magnitude slower in larger directories, however in absolute time this is still quite short. The slowest check (`/mnt/mountpoint/subdirectory`) takes 0.001 seconds which should be insignificant.
- All of the tests returned the same results for all test cases.
- This functionality is opt-in for each call, so will only decrease performance where users choose it (e.g. when checking paths which are more likely to be stuck mounts).

**Testing**
```
/usr/directory/symlink
Method 1: 4624.13
Method 2: 1677.48
Method 3: 5644.01

/usr/bigdirectory/symlink
Method 1: 791032.06
Method 2: 1644.84
Method 3: 8258.53

/usr/directory
Method 1: 9594.34
Method 2: 1570.05
Method 3: 4841.86

/usr/bigdirectory
Method 1: 8438.8
Method 2: 1579.82
Method 3: 4776.39

/usr/directory/broken-symlink
Method 1: 4252.2
Method 2: 1792.68
Method 3: 5062.69

/usr/bigdirectory/broken-symlink
Method 1: 821249.25
Method 2: 1655.62
Method 3: 5294.16

/usr/directory/target
Method 1: 3891.3
Method 2: 1586.6
Method 3: 5027.29

/usr/bigdirectory/target
Method 1: 314027.85
Method 2: 1637.95
Method 3: 5279.84

/usr/directory/non-existant-target
Method 1: 4454.66
Method 2: 1685.95
Method 3: 5353.8

/usr/bigdirectory/non-existant-target
Method 1: 970405.14
Method 2: 1846.52
Method 3: 5508.54

/mnt/mountpoint
Method 1: 19146.12
Method 2: 17161.07
Method 3: 4938.25

/mnt/mountpoint/
Method 1: 1273730.07
Method 2: 1827.36
Method 3: 4839.14

/symlink-targeting-mountpoint
Method 1: 20555.64
Method 2: 1521.84
Method 3: 4234.33

/mnt/mountpoint/subdirectory
Method 1: 1271890.77
Method 2: 19262.63
Method 3: 5596.77

/
Method 1: 28275.14
Method 2: 1443.39
Method 3: 4180.43


```

Reply:
Thank You for the awesome explanation!